CorsixTH/Src/iso_fs.cpp

```diff
@@ -70,33 +75,187 @@ constexpr int max_directory_depth = 16;
 
 /// Reasonably unique name of a file from Theme Hospital that can be used to
 /// indicate that we've loaded the right directory.
-constexpr std::array<const char, 10> vblk_0_filename {'V','B','L','K','-','0','.','T','A','B'};
+constexpr const char*  vblk_0_filename = "VBLK-0.TAB";
 
 /// Sector sizes can vary, but they must be powers of two, and the minimum
 /// size is 2048.
 constexpr size_t min_sector_size = 2048;
 
-template <class T> inline T read_native_int(const uint8_t *p)
+/// Offset of the sector size from the primary volume descriptor
+constexpr size_t sector_size_offset = 128;
+
+/// Offset of the root directory entry from the primary volume descriptor
+constexpr ptrdiff_t root_directory_offset = 156;
+
+/// The root directory entry is a fixed size.
+constexpr size_t root_directory_entry_size = 34;
+
+/// ISO 9660 has a 32kb reserve area at the beginning of the file formal
+/// e.g. boot information.
+constexpr uint32_t first_filesystem_sector = 16;
+
+void trim_identifier_version(const uint8_t* sIdent, uint8_t& iLength)
 {
-    // ISO 9660 commonly encodes multi-byte integers as little endian followed
-    // by big endian. Note that the first byte of iEndianness will be a zero on
-    // little endian systems, and a one on big endian.
-    static const uint16_t iEndianness = 0x0100;
-    return reinterpret_cast<const T*>(p)[*reinterpret_cast<const uint8_t*>(&iEndianness)];
+    for(uint8_t i = 0; i < iLength; ++i)
+    {
+        if(sIdent[i] == ';')
+        {
+            iLength = i;
+            return;
+        }
+    }
 }
 
-} // namespace
+char normalise(char c) {
+    if (c == '_') {
+        return '-';
+    } else if ('a' <= c && c <= 'z') {
+        return static_cast<char>(c - 'a' + 'A');
+    } else {
+        return c;
+    }
+}
 
-iso_filesystem::iso_filesystem()
-{
-    raw_file = nullptr;
-    error = nullptr;
-    files = nullptr;
-    file_count = 0;
-    file_table_size = 0;
-    path_seperator = '\\';
+std::string normalise(const uint8_t* start, size_t length) {
+    std::string ret;
+    const uint8_t* p = start;
+    for (size_t i = 0; i < length; i++) {
+        ret.push_back(normalise(static_cast<char>(*p)));
+        ++p;
+    }
+    return ret;
 }
 
+std::string normalise(const char* str) {
+    std::string ret;
+    const char* p = str;
```

Comment:
You can use `str` instead of introducing `p`, but this is ok too.

Reply:
Yeah, I don't like mutating arguments. Can easily miss if passed by reference or not.